Unify logic to checkout a pull request

rafeca opened the pull request:
Closes #9439
Closes #9593

This PR is based off https://github.com/desktop/desktop/pull/9340

## Description

This PR contains a small refactor to unify the logic to checkout a PR from GitHub.

Currently, there are two ways to checkout a PR:

1. From the "Pull Requests" list tab that appears on the GitHub Desktop app.
    <img width="440" alt="Screenshot 2020-03-27 at 15 42 47" src="https://user-images.githubusercontent.com/408035/77767616-a1ebed00-7041-11ea-811b-ca6f38f22274.png">
2. From github.com when navigating to a Pull Request page:
    <img width="925" alt="Screenshot 2020-03-27 at 15 43 15" src="https://user-images.githubusercontent.com/408035/77767735-c6e06000-7041-11ea-80de-d62e618c95ce.png">

Each of these entry points has its own logic, which causes some inconsistencies between them (as we found in https://github.com/desktop/desktop/pull/9340#issuecomment-603953203).

## Test cases

I've tried to follow the test cases that were written in https://github.com/desktop/desktop/pull/9138, but couldn't check every single one of them.

## Release notes

Notes: no-notes

rafeca commented on app/src/lib/stores/app-store.ts:
```diff
@@ -5271,18 +5272,47 @@ export class AppStore extends TypedBaseStore<IAppState> {
+        cloneURL: pullRequest.head.repo.clone_url,
+        ref: pullRequest.head.ref,
+      }
+    }
```
Not particularly proud of this piece of logic, but it allows the `_checkoutPullRequest()` method to accept both a `PullRequest` object and an `IAPIPullRequest` one.

I can also move this logic into a separate method to try to keep this method cleaner.

rafeca commented on app/src/ui/dispatcher/dispatcher.ts:
```diff
@@ -1400,99 +1404,106 @@ export class Dispatcher {
+      if (this.doesRepositoryMatchUrl(repository, headUrl)) {
+        return repository
+      }
+    }
```
I wanted to use `repositories.find()` here but TypeScript's `Array.prototype.find` is not capable to infer that the found element can only be of type `Repository | null`.

tierninho commented:
Ran through almost all the tests here https://github.com/desktop/desktop/pull/9138#issuecomment-592714953 and only one stood out -- the rest passed.

>  PR in upstream with upstream as source —> opens fork in Desktop and source branch is checked out (not-kuychaco/urban-potato#3)

To set up:
- Fork https://github.com/not-kuychaco/urban-potato, stay on master branch
- Go to any other repo on Desktop
- Go to https://github.com/not-kuychaco/urban-potato/pull/3 and click Open with Desktop link. 

Expected: branch opens in forked repo
Actual: the Clone Repository modal opens

rafeca commented:
>Ran through almost all the tests here #9138 (comment) and only one stood out -- the rest passed.

Thanks! ✨✨

I fixed the issue in the last commit. This was caused because the URL that's opened in that case is `x-github-client://openRepo/https://github.com/not-kuychaco/urban-potato?branch=kuychaco-patch-1`, when it should be something like `x-github-client://openRepo/https://github.com/not-kuychaco/urban-potato?pr=3`. I fixed the issue in Desktop, but I'll try to change the logic in dotcom to make the URLs from all PRs consistent.

niik commented:
> Hey, I was looking to test some of the scenarios here, but is there a merge conflict that needs resolving?

@tierninho I think this is because @rafeca force-pushed the branch at one point

![image](https://user-images.githubusercontent.com/634063/91437900-de2eae00-e86a-11ea-82fd-3256eb820b02.png)

Unfortunately for you it means it's gonna be a bit trickier to get up to date. Either you delete the branch locally and check it out again or you can run the following command in your terminal

```git reset --hard  origin/techdebt/unify-checkout-pr-logic```

rafeca commented:
> @tierninho I think this is because @rafeca force-pushed the branch at one point

Ups, sorry about that @tierninho ! and thanks @niik for figuring it out!

niik reviewed (approved):
Great work cleaning this up @rafeca!
